torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
   return new For(var_new, start_new, stop_new, body_new, loop_options);
 }
 
+// Simplify division using distributive laws for the following cases:
+// 1) (i + 24) / 6, i is the index var of a for-stmt and the range of i is
+// contained in [0, 6)
+//    -> 4
+// 2) (i + j*6) / 6, i, j are index vars of for-stmts and the range of i is
+// contained in [0, 6) + the start value of j is not less than 0
+//    -> j
+const Expr* distributeDiv(
+    const Expr* lhs,
+    const Expr* rhs,
+    VarBoundInfo var_bound_info) {
+  if (!lhs || !rhs) {
+    return nullptr;
+  }
+  auto* lhsAdd = dynamic_cast<const Add*>(lhs);
+  const Expr* rhsScalar = rhs->isConstant() ? rhs : nullptr;
+
+  if (!lhsAdd || !rhsScalar) {
+    return nullptr;
+  }
+  const Expr* lhsAdd1 = lhsAdd->lhs();
+  const Expr* lhsAdd2 = lhsAdd->rhs();
+
+  // identify index var 'i'
+  const Var* var_key = dynamic_cast<const Var*>(lhsAdd1);
+  const Expr* main = lhsAdd2;
+  if (var_key == nullptr) {
+    var_key = dynamic_cast<const Var*>(lhsAdd2);
+    main = lhsAdd1;
+  }
+
+  if (var_key != nullptr) {
+    auto got = var_bound_info.find(var_key);
+    if (got == var_bound_info.end()) {
+      return nullptr;
+    }
+
+    // check the bounds of 'i'
+    auto start = got->second.first;
+    auto end = got->second.second;
+    const Expr* check_start =
+        IRSimplifier::simplify(new CompareSelect(start, new IntImm(0), kGE));
+    const Expr* check_end =
+        IRSimplifier::simplify(new CompareSelect(end, rhsScalar, kLE));
+    if (check_start->isConstant() && check_end->isConstant() &&
+        immediateEquals(check_start, 1) && immediateEquals(check_end, 1)) {
+      const Expr* ret = IRSimplifier::simplify(new Div(main, rhsScalar));
+
+      // simplify type 1 exprs: '(i+24)/6'
+      const Expr* mod_check = IRSimplifier::simplify(new Mod(main, rhsScalar));
+      if (mod_check->isConstant() && immediateEquals(mod_check, 0)) {
+        return ret;
+      }
+
+      // simplify type 2 exprs: '(i+j*6)/6'
+      auto ret_var = dynamic_cast<const Var*>(ret);
+      if (ret_var) {
+        auto got = var_bound_info.find(ret_var);
+        if (got == var_bound_info.end()) {
+          return nullptr;
+        }
+        check_start = IRSimplifier::simplify(
+            new CompareSelect(start, new IntImm(0), kGE));
+        if (check_start->isConstant() && immediateEquals(check_start, 1)) {
+          return ret_var;
+        }
+      }
+    }
+  }
+  return nullptr;
+}
+
+// Simplify mod using distributive laws for the following cases:
+// 1) (i + 24) % 6, i is the index var of a for-stmt and the range of i is
```

Comment:
I wonder if the implementation would be simpler if we broke the optimization into two steps:
1) `(x + y) % CONST` --> `x % CONST + y % CONST`
2) `x % CONST -> x` if `0 <= x < CONST`

Similarly for div it would be:
1) `(x + y) / CONST` --> `x / CONST + y / CONST` if `x % CONST == 0` or `y % CONST == 0`
2) `x / CONST -> 0` if `0 <= x < CONST`

Here only the second step would be using range info. One downside of this approach is that we could distribute divs/mods without the subsequent optimization, but I wonder what you think?

Reply:
Btw, #52580 is related to this. Maybe we should fix it too while we're at it.

Reply:
I like this idea! I already have a PR for #52580: PR #60693 but it is not a clean solution to me. I was wondering how to combine the two more neatly and your suggestion seems to be the right way to go. 

Basically, in the first step we distribute divs/mods if `x % CONST == 0 or y % CONST == 0` and perform and the range based simplification in second step. The first step is also a simplification even without the subsequent optimization: 
`(i + j*6)%6 --> i%6` or `(i + j*6)/6 --> i/6 + j`

After the first step, the mod expr is still a mod which should not cause any trouble. The distribution of div is also a good thing? It simplified the expression into a polynomial-like expression for subsequent simplifications on polynomials?

Reply:
Discussed this offline and agreed that having a simplification with the two steps combined along with a separate simplification for (2) is a way to go, because (1) needs a range info anyway.